File: tests/e2e/pty-drive.py
#!/usr/bin/env python3
"""Drive an interactive api-key-case command under a real pty.

Why this exists: packages/cli/prompt.ts refuses to read a secret value or a
production confirmation from anything that is not a TTY, and AGENTS.md
section 3 forbids adding a bypass ("there is no flag to skip this").  So the
end-to-end job must not reach for a --stdin escape hatch; it has to type into
a real terminal the way a human does.  This allocates a pty, waits for each
prompt to actually appear, and types the answer.

Secret material is passed by environment variable name, never on argv, and is
masked out of both the printed transcript and the transcript file.  Literal
step answers ("yes") are deliberately NOT masked: they are not secrets, and
masking such common words would corrupt the transcript the caller asserts on.

Usage:
  pty-drive.py --step 'PATTERN=>env:VARNAME' \
               --step 'PATTERN=>literal:yes' \
               [--mask-env VARNAME] [--transcript FILE] [--timeout SECONDS] \
               -- command arg...

Exit status:
  the child's exit status, or
  97  a --step pattern never appeared before the child exited
  98  the overall timeout elapsed
"""

import argparse
import errno
import os
import pty
import re
import select
import subprocess
import sys
import termios
import time

# Once the child is gone, keep draining the pty briefly so the last lines it
# wrote before exiting still make it into the transcript.
DRAIN_SECONDS = 2.0


def parse_step(raw):
    """'PATTERN=>env:NAME' or 'PATTERN=>literal:TEXT' -> (compiled, text, secret)."""
    if "=>" not in raw:
        raise argparse.ArgumentTypeError(f"--step needs PATTERN=>SOURCE:VALUE, got {raw!r}")
    pattern, source = raw.split("=>", 1)
    if source.startswith("env:"):
        name = source[len("env:"):]
        value = os.environ.get(name)
        if not value:
            raise argparse.ArgumentTypeError(f"--step referenced env {name}, which is unset or empty")
        secret = True
    elif source.startswith("literal:"):
        value = source[len("literal:"):]
        secret = False
    else:
        raise argparse.ArgumentTypeError(f"--step source must be env: or literal:, got {source!r}")
    return re.compile(pattern), value, secret


def mask(text, secrets):
    for secret in secrets:
        if secret:
            text = text.replace(secret, "***MASKED***")
    return text


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--step", action="append", default=[], type=parse_step)
    parser.add_argument("--mask-env", action="append", default=[])
    parser.add_argument("--transcript")
    # Unmasked copy, for the caller's leak assertion. Grepping the masked
    # transcript would be circular: the mask would hide the very leak the
    # check exists to find. Callers must delete it and must never print it.
    parser.add_argument("--raw-transcript")
    parser.add_argument("--timeout", type=float, default=240.0)
    parser.add_argument("command", nargs=argparse.REMAINDER)
    args = parser.parse_args()

    command = args.command[1:] if args.command and args.command[0] == "--" else args.command
    if not command:
        parser.error("no command given (put it after --)")

    secrets = [os.environ.get(name, "") for name in args.mask_env]
    secrets += [value for _pattern, value, is_secret in args.step if is_secret]

    master_fd, slave_fd = pty.openpty()

    # Turn the terminal's own echo off before the child starts.
    #
    # A pty's line discipline echoes whatever is typed at it straight back into
    # the transcript. For a human that never matters here: prompt.ts prints the
    # prompt and only then calls readHiddenLine(), which switches the tty to raw
    # mode (echo off), and a person takes far longer to reach for a key than
    # that takes. This harness answers within microseconds of seeing the prompt,
    # so it wins that race and its own keystrokes land in the transcript --
    # which the leak check would then report as the product having printed the
    # secret. Disabling echo up front removes the artifact without hiding
    # anything real: the CLI's own writes are unaffected, and node's readline
    # renders the confirmation answer itself rather than relying on tty echo.
    attributes = termios.tcgetattr(slave_fd)
    attributes[3] &= ~termios.ECHO  # lflag
    termios.tcsetattr(slave_fd, termios.TCSANOW, attributes)

    child = subprocess.Popen(
        command,
        stdin=slave_fd,
        stdout=slave_fd,
        stderr=slave_fd,
        close_fds=True,
        start_new_session=True,
    )
    os.close(slave_fd)

    transcript = ""
    # Only text seen since the last answer is matched, so one prompt cannot be
    # satisfied twice by text still sitting in the buffer.
    pending = ""
    steps = list(args.step)
    deadline = time.monotonic() + args.timeout
    drain_until = None
    timed_out = False

    try:
        while True:
            now = time.monotonic()
            if now > deadline:
                timed_out = True
                child.kill()
                break
            if drain_until is not None and now > drain_until:
                break

            try:
                ready, _, _ = select.select([master_fd], [], [], 0.25)
            except (OSError, ValueError):
                break

            if not ready:
                if drain_until is None and child.poll() is not None:
                    drain_until = time.monotonic() + DRAIN_SECONDS
                continue

            try:
                chunk = os.read(master_fd, 4096)
            except OSError as err:
                # The child closed the pty: normal end of stream on macOS/Linux.
                if err.errno in (errno.EIO, errno.EBADF):
                    break
                raise
            if not chunk:
                break

            text = chunk.decode("utf-8", "replace")
            transcript += text
            pending += text

            if steps and steps[0][0].search(pending):
                _pattern, value, _secret = steps.pop(0)
                os.write(master_fd, (value + "\r").encode("utf-8"))
                pending = ""
    finally:
        try:
            os.close(master_fd)
        except OSError:
            pass

    exit_code = child.wait()
    masked = mask(transcript, secrets)

    if args.transcript:
        with open(args.transcript, "w", encoding="utf-8") as handle:
            handle.write(masked)

    if args.raw_transcript:
        fd = os.open(args.raw_transcript, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(transcript)

    sys.stdout.write(masked)
    if not masked.endswith("\n"):
        sys.stdout.write("\n")
    sys.stdout.flush()

    if timed_out:
        sys.stderr.write(f"pty-drive: timed out after {args.timeout}s\n")
        return 98
    if steps:
        missing = ", ".join(pattern.pattern for pattern, _value, _secret in steps)
        sys.stderr.write(f"pty-drive: child exited before these prompts appeared: {missing}\n")
        return 97
    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: tests/e2e/pty-drive.mjs
#!/usr/bin/env node
// Windows counterpart of pty-drive.py, driving a ConPTY instead of a POSIX pty.
//
// Same reason for existing: packages/cli/prompt.ts refuses to read a secret
// value or a production confirmation from anything that is not a TTY, and
// AGENTS.md section 3 forbids adding a bypass ("there is no flag to skip
// this"). So the end-to-end job types into a real console the way a human
// does, rather than weakening the product to make it testable.
//
// ConPTY (Windows 10 1809+) is what gives the child a genuine console handle,
// so `process.stdin.isTTY` is true inside it and the hidden-input path runs
// exactly as it does for a user. node-pty is installed by the e2e workflow
// only -- it is never a dependency of the published package.
//
// Same command-line contract as pty-drive.py, so tests/e2e/deploy-e2e.sh can
// swap one for the other by platform.
//
// Usage:
//   node pty-drive.mjs --step 'PROMPT=>env:VARNAME' \
//                      --step 'PROMPT=>literal:yes' \
//                      [--mask-env VARNAME] [--transcript FILE] \
//                      [--raw-transcript FILE] [--timeout SECONDS] \
//                      -- command arg...
//
// Exit status: the child's, or 97 (a prompt never appeared) / 98 (timeout).

import { existsSync, openSync, writeFileSync, writeSync, closeSync } from "node:fs";
import { delimiter, isAbsolute, join } from "node:path";

// A human takes a moment to reach for the keyboard; this harness does not.
// prompt.ts writes the prompt and only then switches the console to raw mode
// (echo off), so answering instantly can land our own keystrokes in the
// transcript as console echo -- indistinguishable, to the leak check, from
// the product having printed the value. Pausing bridges that window. The
// POSIX driver solves the same problem by clearing ECHO on the pty, which
// ConPTY does not expose.
const ANSWER_DELAY_MS = 250;

const ESC = String.fromCharCode(0x1b);
const BEL = String.fromCharCode(0x07);

function parseStep(raw) {
  const marker = raw.indexOf("=>");
  if (marker === -1) {
    throw new Error(`--step needs PROMPT=>SOURCE:VALUE, got ${JSON.stringify(raw)}`);
  }
  const prompt = raw.slice(0, marker);
  const source = raw.slice(marker + 2);
  if (!prompt) {
    throw new Error("--step prompt must not be empty");
  }

  if (source.startsWith("env:")) {
    const name = source.slice(4);
    const value = process.env[name];
    if (!value) {
      throw new Error(`--step referenced env ${name}, which is unset or empty`);
    }
    return { prompt, value, secret: true };
  }
  if (source.startsWith("literal:")) {
    // Literal answers ("yes") are deliberately not masked: they are not
    // secrets, and masking such common words would corrupt the transcript
    // the caller asserts on.
    return { prompt, value: source.slice(8), secret: false };
  }
  throw new Error(`--step source must be env: or literal:, got ${JSON.stringify(source)}`);
}

function parseArgs(argv) {
  const opts = {
    steps: [],
    maskEnv: [],
    transcript: null,
    rawTranscript: null,
    timeout: 240,
    command: []
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--step") opts.steps.push(parseStep(argv[++i]));
    else if (arg === "--mask-env") opts.maskEnv.push(argv[++i]);
    else if (arg === "--transcript") opts.transcript = argv[++i];
    else if (arg === "--raw-transcript") opts.rawTranscript = argv[++i];
    else if (arg === "--timeout") opts.timeout = Number(argv[++i]);
    else if (arg === "--") {
      opts.command = argv.slice(i + 1);
      break;
    } else {
      opts.command = argv.slice(i);
      break;
    }
  }
  if (opts.command.length === 0) throw new Error("no command given (put it after --)");
  return opts;
}

// ConPTY repaints the screen: cursor moves, clears and colour runs are mixed
// into the byte stream. Strip them so the caller's greps see plain text.
// This removes escape sequences only -- it can never remove a secret value,
// so the leak check (which reads the raw transcript) stays honest.
const OSC = new RegExp(`${ESC}\\][\\s\\S]*?(?:${BEL}|${ESC}\\\\)`, "g");
const CSI = new RegExp(`${ESC}\\[[0-9;?]*[ -/]*[@-~]`, "g");
const OTHER_ESCAPE = new RegExp(`${ESC}[@-Z\\\\-_]`, "g");

function stripAnsi(text) {
  return text
    .replace(OSC, "")
    .replace(CSI, "")
    .replace(OTHER_ESCAPE, "")
    .replace(/\r(?!\n)/g, ""); // bare CR from repaints; CRLF stays a line break
}

function mask(text, secrets) {
  let out = text;
  for (const secret of secrets) {
    if (secret) out = out.split(secret).join("***MASKED***");
  }
  return out;
}

function writePrivate(path, contents) {
  const fd = openSync(path, "w", 0o600);
  try {
    writeSync(fd, contents);
  } finally {
    closeSync(fd);
  }
}

const opts = parseArgs(process.argv.slice(2));

let pty;
try {
  pty = await import("node-pty");
} catch {
  process.stderr.write(
    "pty-drive: node-pty is not installed. The Windows e2e job installs it with\n" +
      "  npm install --no-save node-pty\n" +
      "It is intentionally not a dependency of the published package.\n"
  );
  process.exit(2);
}

const secrets = [
  ...opts.maskEnv.map((name) => process.env[name] ?? ""),
  ...opts.steps.filter((step) => step.secret).map((step) => step.value)
];

// node-pty's Windows agent needs a real path -- it does no PATH lookup of its
// own and fails with "File not found" on a bare command name. This resolves
// the harness's own launcher (always `node`), and has nothing to do with the
// product's deliberately PATH-only vendor-CLI resolution in
// packages/core/deploy/which.ts.
function resolveCommand(command) {
  if (isAbsolute(command) || command.includes("/") || command.includes("\\")) return command;
  if (command === "node" || command === "node.exe") return process.execPath;

  const extensions = process.platform === "win32" ? [".exe", ".cmd", ".bat", ""] : [""];
  for (const dir of (process.env.PATH ?? "").split(delimiter).filter(Boolean)) {
    for (const extension of extensions) {
      const candidate = join(dir, command + extension);
      if (existsSync(candidate)) return candidate;
    }
  }
  throw new Error(`could not resolve ${command} from PATH`);
}

const [command, ...args] = opts.command;
const child = pty.spawn(resolveCommand(command), args, {
  name: "xterm-256color",
  // Wide enough that ConPTY does not hard-wrap a line mid-value, which would
  // split a leaked secret across a newline and hide it from a plain grep.
  cols: 400,
  rows: 60,
  cwd: process.cwd(),
  env: process.env,
  useConpty: true
});

let transcript = "";
// Only text seen since the last answer is matched, so one prompt cannot be
// satisfied twice by text still sitting in the buffer.
let pending = "";
const steps = [...opts.steps];
let answering = false;
let settled = false;

function finish(code, reason) {
  if (settled) return;
  settled = true;
  clearTimeout(timer);

  // Collapse the runs of blank lines ConPTY's repaints leave behind once the
  // cursor-movement escapes are gone. Display copy only -- the raw transcript
  // the leak check reads is never touched.
  const masked = mask(stripAnsi(transcript), secrets).replace(/\n{3,}/g, "\n\n");

  if (opts.transcript) writeFileSync(opts.transcript, masked, "utf8");
  // Unmasked copy for the caller's leak assertion. Grepping the masked
  // transcript would be circular: the mask would hide the very leak the
  // check exists to find. Never printed; deleted by the caller.
  if (opts.rawTranscript) writePrivate(opts.rawTranscript, transcript);

  process.stdout.write(masked.endsWith("\n") ? masked : `${masked}\n`);

  if (reason === "timeout") {
    process.stderr.write(`pty-drive: timed out after ${opts.timeout}s\n`);
    process.exit(98);
  }
  if (steps.length > 0) {
    const missing = steps.map((step) => step.prompt).join(", ");
    process.stderr.write(`pty-drive: child exited before these prompts appeared: ${missing}\n`);
    process.exit(97);
  }
  // node-pty keeps the event loop alive on its own, so exit explicitly.
  process.exit(code ?? 1);
}

const timer = setTimeout(() => {
  try {
    child.kill();
  } catch {
    /* already gone */
  }
  finish(null, "timeout");
}, opts.timeout * 1000);

child.onData((chunk) => {
  transcript += chunk;
  pending += chunk;

  if (answering || steps.length === 0) return;
  if (!stripAnsi(pending).includes(steps[0].prompt)) return;

  answering = true;
  const { value } = steps.shift();
  setTimeout(() => {
    child.write(`${value}\r`);
    pending = "";
    answering = false;
  }, ANSWER_DELAY_MS);
});

child.onExit(({ exitCode }) => {
  // Let any final bytes written just before exit arrive before settling.
  setTimeout(() => finish(exitCode, "exit"), 300);
});
